src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities;
+
+internal static class AsyncLazy
+{
+    public static AsyncLazy<T, TArg> Create<T, TArg>(Func<TArg, CancellationToken, Task<T>> asynchronousComputeFunction, TArg arg)
+        => new AsyncLazy<T, TArg>(asynchronousComputeFunction, arg);
+
+    public static AsyncLazy<T, TArg> Create<T, TArg>(Func<TArg, CancellationToken, T> synchronousComputeFunction, TArg arg)
+        => new AsyncLazy<T, TArg>((arg, cancellationToken) => Task.FromResult(synchronousComputeFunction(arg, cancellationToken)), synchronousComputeFunction, arg);
+
+    public static AsyncLazy<T, TArg> Create<T, TArg>(Func<TArg, CancellationToken, Task<T>> asynchronousComputeFunction, Func<TArg, CancellationToken, T> synchronousComputeFunction, TArg arg)
+        => new AsyncLazy<T, TArg>(asynchronousComputeFunction, synchronousComputeFunction, arg);
+
+    public static AsyncLazy<T> Create<T>(Func<CancellationToken, Task<T>> asynchronousComputeFunction)
+        => new AsyncLazyWithoutArg<T>(asynchronousComputeFunction);
+
+    public static AsyncLazy<T> Create<T>(Func<CancellationToken, T> synchronousComputeFunction)
+        => new AsyncLazyWithoutArg<T>(cancellationToken => Task.FromResult(synchronousComputeFunction(cancellationToken)), synchronousComputeFunction);
```

Comment:
💡 This allocation falls into the same category as the one I mention above, but is easier to avoid:

```suggestion
    public static AsyncLazy<T> Create<T>(Func<CancellationToken, T> synchronousComputeFunction)
        => Create(
            asynchronousComputeFunction: static (synchronousComputeFunction, cancellationToken) => Task.FromResult(synchronousComputeFunction(cancellationToken)), arg: synchronousComputeFunction);
```